File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/util/gvf/GVF.java
package org.firstinspires.ftc.teamcode.util.gvf;

import org.firstinspires.ftc.teamcode.util.geometry.Point;
import org.firstinspires.ftc.teamcode.util.geometry.Vector2d;
import org.firstinspires.ftc.teamcode.util.path.Path;

public class GVF {
    // TODO: add ability for path sequences
    Path path;
    double kN;

    public GVF(Path path, double kN) {
        this.path = path;
        this.kN = kN;
    }

    public Vector2d getVector(Point p) {
        return path.getTau(p)
                .add(path.getError(p).scale(kN))
                .normalize();
    }
}

File: quackerama/src/main/kotlin/com/scrapmetal/quackerama/control/gvf/GVF.kt
package com.scrapmetal.quackerama.control.gvf

import com.scrapmetal.quackerama.control.Vector2d
import com.scrapmetal.quackerama.control.path.Path

class GVF(val path: Path, val kN: Double) {
    // TODO: add ability for path sequences

    fun getVector(p: Vector2d): Vector2d {
        // TODO: figure out how much to push closestT
        if (path.paths[0].closestT(p) > 0.99) {
            return path.update(p).second.unit
        } else {
            return path.update(p).let { it.first + it.second * kN }.unit
        }
    }
    fun getClosestT(p: Vector2d): Double {
        return path.paths[0].closestT(p)
    }
}